### DISCUSSION TITLE
authrntication with e-mail

### DISCUSSION THREAD

**Discussion by dongri727** [serverpod/serverpod#1985] [Q&A]
I am trying to perform email authentication.

serverpod has been deployed to AWS.
All related modules etc. have been unified to 1.2.5.

The front end is currently hosted on firebase.
(I can't get it hosted in AWS properly, I'll ask about this separately).

When I try to register a new email, I get an "E-mail already in use" error.

When I change the environment to local Docker and test,
I receive a confirmation email to the new address,
When I enter the Verify Code, it is registered.

What is the difference from the production environment?
What should I check?

Translated with www.DeepL.com/Translator (free version)

**Comment by SandPod** [comment]
Hey,

How are you doing your authentication process?
Could you paste any logs that are produced?

I'm not familiar with a "E-mail already in use" error in the mail authentication provider.

**Comment by SandPod** [comment]
I have found the error message that you are encountering now. There are a couple of reasons this might be displayed and I have documented them in this post here: https://github.com/serverpod/serverpod/issues/1969#issuecomment-1968400800

**Comment by dongri727** [comment]
I have read the above Issue.
Based on that, I will describe my situation in a little more detail.

The email_create_request in the database records the userName, email, hash, and verificationCode that the testers tried to register.
However, the testers did not receive the Code (we also checked the spam mail).
The display is as shown in the attachment.
Of course, it is not registered in user_info.

What I don't understand is why it works properly when I am using Docker in my local environment,

I don't understand why it works fine when using Docker locally, but not on AWS.
Is there something I need to change in my configuration?

<img width="346" alt="スクリーンショット 2024-03-04 7 05 58" src="https://github.com/serverpod/serverpod/assets/84567162/ecfa90bd-0a54-4a50-ae0c-6e9cc0f4cefb">

**Comment by dongri727** [comment]
I have a feeling that it might be the mailer. I will check it out.

**Reply by SandPod** [reply]
Given the description above this sounds like the root of the problem. A strong contender is also that the mailar can't be reached from the AWS environment. Is it possible that there are firewalls in place that intercept the call?

Since you can find the record in the database I'm confident that the `sendValidationEmail` callback is called (you could place a log statement there just to ensure this).

**Comment by dongri727** [comment]
My mailer is still gmail,
Changed the app password,　
Modified passwords.yaml,
SERVERPOD_PASSWORDS in Repository secrets in Git Hub and,
and deployed
The situation is still the same.
Is there anything else that needs to be updated?

**Reply by SandPod** [reply]
Have you validated that the sendValidationEmail callback is called on the server and that the request reaches the mailer?

I would recommend looking at the steps in our article on how to get email authentication working: https://medium.com/serverpod/getting-started-with-serverpod-authentication-part-1-72c25280e6e9

**Reply by dongri727** [reply]
Do I need to do anything after I edited SERVERPOD_PASSWORDS on GitHub ? Like terraform unit or apply before deploy AWS?

**Reply by SandPod** [reply]
As far as I know, more modifications shouldn't be required since the Github action takes the password file and writes it into the docker container it builds. If the passwords were not there you shouldn't have been able to connect to the database.

I think it would be good to inspect the response from your mailer when you send the validation email. If this is not a successful response it might contain more information about what is going wrong.

**Reply by SandPod** [reply]
One thought: in the Github repository secret `SERVERPOD_PASSWORDS`, did you add the `gmailPassword` and `gmailEmail` to both the production and development environment?

**Comment by dongri727** [comment]
I have successfully used Gmail for email services in a local environment with Serverpod, but encountered issues when trying to replicate this setup on AWS. Therefore, I've decided to switch to AWS Simple Email Service (SES). While the SES setup is complete, I am uncertain about the necessary adjustments on the Serverpod side.

Could you please confirm if the following changes are correct for integrating SES?

Should the gmailEmail parameter be replaced with sesEmail? If so, does this change apply directly within any specific configuration file or code snippet that previously handled Gmail settings?
The main concern arises with handling what was formerly gmailPassword. I understand this might now involve using AWS keys. Should this key be specified in place of gmailPassword within the passwords.yaml file, or is there an alternative approach for securely integrating AWS keys with Serverpod that does not require modifying this file?
Lastly, am I missing any critical steps or settings unique to SES that need to be addressed to ensure seamless integration with Serverpod?

Thank you for your assistance.

**Reply by SandPod** [reply]
You are free to store any secret under any name in the `SERVERPOD_PASSWORDS` Github repository secret or the `passwords.yaml` file in your project.

In the code, you then fetch that password based on the name you picked.
If I create a secret named `alexSecret` I would fetch the value of that secret using `session.serverpod.getPassword('alexSecret')`.

Note that you need to ensure that the environment you are running the application in has access to the secret. (When deploying to AWS the server is run in the `production` environment.

Secrets placed in the `shared` namespace will be available to all environments.

This makes it possible to configure different mailers depending on what secrets are available, in a prod environment it is good to run another service than the Gmail mailer for example.

So to move to SES just ensure you have defined the secrets required to get it running locally, but in the `production` environment in your `SERVERPOD_PASSWORDS` Github repository secret and you should be good to go.
